db/migrate/20231210061330_add_created_at_to_fts_targets_with_index.rb

```diff
@@ -0,0 +1,109 @@
+  def up
+    return if !table_exists?(:fts_targets)
+
+    add_column :fts_targets, :created_at, :timestamp
+
+    ActiveRecord::Base.transaction do
+      update_created_at_using_last_modified_at
+      update_created_at_using_created_on
+    end
+
+    if Redmine::Database.mysql?
+      add_index :fts_targets, :created_at
+    else
+      remove_index :fts_targets, name: "fts_targets_index_pgroonga"
+      add_index :fts_targets,
+                [:id,
+                 :source_id,
+                 :source_type_id,
+                 :project_id,
+                 :container_id,
+                 :container_type_id,
+                 :custom_field_id,
+                 :is_private,
+                 :last_modified_at,
+                 :created_at,
+                 :title,
+                 :content,
+                 :tag_ids],
+                using: "PGroonga",
+                with: "normalizer = 'NormalizerNFKC121'",
+                name: "fts_targets_index_pgroonga"
+    end
+  end
+
+  def down
+    return if !table_exists?(:fts_targets)
+
+    if Redmine::Database.mysql?
+      remove_index :fts_targets, :created_at
+    else
+      remove_index :fts_targets, name: "fts_targets_index_pgroonga"
+      add_index :fts_targets,
+                [:id,
+                 :source_id,
+                 :source_type_id,
+                 :project_id,
+                 :container_id,
+                 :container_type_id,
+                 :custom_field_id,
+                 :is_private,
+                 :last_modified_at,
+                 :title,
+                 :content,
+                 :tag_ids],
+                using: "PGroonga",
+                with: "normalizer = 'NormalizerNFKC121'",
+                name: "fts_targets_index_pgroonga"
+    end
+
+    remove_column :fts_targets, :created_at
+  end
+
+  private
+
+  class FtsTarget < ActiveRecord::Base
+  end
+
+  class FtsType < ActiveRecord::Base
+  end
+
+  class Issue < ActiveRecord::Base
+  end
+
+  class Message < ActiveRecord::Base
+  end
+
+  class Project < ActiveRecord::Base
+  end
+
+  class WikiPage < ActiveRecord::Base
+  end
+
+  def update_created_at_using_last_modified_at
+    types = ['Attachment', 'Change', 'Changeset', 'CustomValue', 'Document', 'Journal', 'News']
+
+    FtsTarget.where(source_type_id: FtsType.where(name: types).select(:id))
+             .update_all(created_at: FtsTarget.arel_table[:last_modified_at])
+  end
+
+  def update_created_at_using_created_on
+    types = ['Issue', 'Message', 'Project', 'WikiPage']
+
+    types.each do |type_name|
+      source_table = type_name.constantize.arel_table
+      fts_targets_table = FtsTarget.arel_table
+
+      subquery_for_created_on = source_table.project(source_table[:created_on])
+                                            .where(source_table[:id].eq(fts_targets_table[:source_id]))
+                                            .to_sql
```

Comment:
What SQL is generated by this?

Can we just use direct Arel only in `where` something like `Issue.where(Issue.arel_table[:id].eq(FtsTarget.arel_table[:source_id])).select(:created_on).to_sql`?

Reply:
fix: [0f8a828](https://github.com/clear-code/redmine_full_text_search/pull/118/commits/0f8a8289c3bb7e9c16f1a984208a95dc6bdd9853)

> What SQL is generated by this?

Following logs is the generated queries.
```console
> subquery_for_created_on
=> "SELECT \"issues\".\"created_on\" FROM \"issues\" WHERE \"issues\".\"id\" = \"fts_targets\".\"source_id\""
```

```console
rb(main):026:0> types = [Issue, Message, Project, WikiPage]
=> [Issue(id: integer, tracker_id: integer, project_id: integer, subject: string, description: text, due_date: date, cate...
irb(main):027:1* types.each do |type|
irb(main):028:1*       subquery_for_created_on = type.where(type.arel_table[:id].eq(FtsTarget.arel_table[:source_id]))
irb(main):029:1>                                     .select(:created_on)
irb(main):030:1>                                     .to_sql
irb(main):031:1>       FtsTarget.where(source_type_id: FtsType.where(name: type.table_name.singularize.camelize).select(:id))
irb(main):032:1>                .update_all("registered_at = (#{subquery_for_created_on})")
irb(main):033:0> end
  FtsTarget Update All (4.2ms)  UPDATE "fts_targets" SET registered_at = (SELECT "issues"."created_on" FROM "issues" WHERE "issues"."id" = "fts_targets"."source_id") WHERE "fts_targets"."source_type_id" IN (SELECT "fts_types"."id" FROM "fts_types" WHERE "fts_types"."name" = $1)  [["name", "Issue"]]
  FtsTarget Update All (2.4ms)  UPDATE "fts_targets" SET registered_at = (SELECT "messages"."created_on" FROM "messages" WHERE "messages"."id" = "fts_targets"."source_id") WHERE "fts_targets"."source_type_id" IN (SELECT "fts_types"."id" FROM "fts_types" WHERE "fts_types"."name" = $1)  [["name", "Message"]]
  FtsTarget Update All (1.8ms)  UPDATE "fts_targets" SET registered_at = (SELECT "projects"."created_on" FROM "projects" WHERE "projects"."id" = "fts_targets"."source_id") WHERE "fts_targets"."source_type_id" IN (SELECT "fts_types"."id" FROM "fts_types" WHERE "fts_types"."name" = $1)  [["name", "Project"]]
  FtsTarget Update All (1.6ms)  UPDATE "fts_targets" SET registered_at = (SELECT "wiki_pages"."created_on" FROM "wiki_pages" WHERE "wiki_pages"."id" = "fts_targets"."source_id") WHERE "fts_targets"."source_type_id" IN (SELECT "fts_types"."id" FROM "fts_types" WHERE "fts_types"."name" = $1)  [["name", "WikiPage"]]
```

Reply:
Thanks.
It seems that the generated SQL is expected.